File: src/channels/registry.ts
import {
  Channel,
  OnInboundMessage,
  OnChatMetadata,
  RegisteredGroup,
} from '../types.js';

export interface ChannelOpts {
  onMessage: OnInboundMessage;
  onChatMetadata: OnChatMetadata;
  registeredGroups: () => Record<string, RegisteredGroup>;
  /**
   * Optional callback for channels to auto-register unregistered groups.
   * Called when a channel receives a message from an unregistered chat.
   * If returned true, the group was registered and message should be processed.
   */
  autoRegisterGroup?: (jid: string, name: string, channel: string) => boolean;
}

export type ChannelFactory = (opts: ChannelOpts) => Channel | null;

const registry = new Map<string, ChannelFactory>();

export function registerChannel(name: string, factory: ChannelFactory): void {
  registry.set(name, factory);
}

export function getChannelFactory(name: string): ChannelFactory | undefined {
  return registry.get(name);
}

export function getRegisteredChannelNames(): string[] {
  return [...registry.keys()];
}

File: iflowclaw/channels/registry.py
from __future__ import annotations

from collections.abc import Callable
from typing import Any

from ..types import Channel, OnChatMetadata, OnInboundMessage, RegisteredGroup

ChannelOpts = dict[str, Any]
ChannelFactory = Callable[[ChannelOpts], Channel | None]

_factories: dict[str, ChannelFactory] = {}


def register_channel(name: str, factory: ChannelFactory) -> None:
    _factories[name] = factory


def get_registered_channel_names() -> list[str]:
    return sorted(_factories.keys())


def get_channel_factory(name: str) -> ChannelFactory | None:
    return _factories.get(name)
